wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/GitCheckoutHiEventDetector.java

```diff
@@ -0,0 +1,114 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2016 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.vfs.impl.file.event;
+
+import com.google.common.annotations.Beta;
+
+import org.eclipse.che.api.core.ForbiddenException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.project.shared.dto.event.GitBranchCheckoutEventDto;
+import org.eclipse.che.api.vfs.Path;
+import org.eclipse.che.api.vfs.VirtualFileSystemProvider;
+import org.slf4j.Logger;
+
+import javax.inject.Inject;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import static java.io.File.separator;
+import static java.util.Optional.empty;
+import static java.util.regex.Pattern.compile;
+import static org.eclipse.che.api.project.shared.dto.event.FileWatcherEventType.MODIFIED;
+import static org.eclipse.che.api.vfs.impl.file.event.HiEvent.Category.UNDEFINED;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Detects if there was a .git/HEAD file modification, which is a sign of git branch
+ * checkout operation, though in some rare cases it simply shows that the head of
```

Comment:
HEAD shows reference to current state of working directory. It can be commit or branch/tag (some reference to commit )